### PR TITLE
Siamese Layer - Complete fix + tests

### PR DESCRIPTION
**Pull request by farizrahman4u**
In this PR:
* Fix `Siamese` layer
* Fix bugs in `__call__` (required by `Siamese`). This includes:
1)  mask support
2) support for nested models
3) temporarily disable layer cache during `__call__`

* Test for `__call__` with nested model
* **Tests for `Siamese` layer (all 5 APIs - 2 Sequential + 3  Graph)**

### CODE REVIEW

**Comment by farizrahman4u** [comment]
@fchollet

**Comment by farizrahman4u** [comment]
@fchollet Some unrelated tests are failing.

**Comment by farizrahman4u** [comment]
@fchollet All set.

**Comment by fchollet** [comment]
Two tests are failing. One is just flaky, but the other is a proper bug, caused by the introduction of `cache_enabled`. Can you look into it?

**Comment by farizrahman4u** [comment]
Oops! I thought both were flaky. Just a sec..

**Comment by farizrahman4u** [comment]
@fchollet Ok, Done.

File: keras/layers/core.py [hunk 1]
```diff
@@ -1444,7 +1465,7 @@ class Siamese(Layer):
         dot_axes: Same meaning as `dot_axes` argument of Merge layer
     '''
     def __init__(self, layer, inputs, merge_mode='concat',
-                 concat_axis=1, dot_axes=-1):
+                 concat_axis=1, dot_axes=-1, is_graph=False):
```
Comment (fchollet) [inline keras/layers/core.py]: Docstring needs to be updated to explain the meaning of `is_graph`.

**Comment by farizrahman4u** [comment]
@fchollet Done

**Comment by farizrahman4u** [comment]
@fchollet Is there anything else?

**Comment by fchollet** [comment]
@farizrahman4u we're good, thank you!

**Pull request merged**